File: 04.Simulation模拟/leetcode_3546_canPartitionGrid.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
    long long tot;
    bool calc(vector<long long>& v){
        long long s = 0;
        for (int i : v){
            s += i;
            if (s*2 == tot){
                return true;
            }
        }
        return false;
    }
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<long long> rows(m, 0), cols(n, 0);
        tot = 0;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                tot += grid[i][j];
                rows[i] += grid[i][j];
                cols[j] += grid[i][j];
            }
        }
        return calc(rows) | calc(cols);
    }
};

int main(){
    Solution s;
    vector<vector<int>> grid = {{1,4},{2,3}};
    cout << s.canPartitionGrid(grid) << endl;
}

File: 04.Simulation模拟/leetcode_3546_canPartitionGrid.py
from typing import List

class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        tot = sum(sum(r) for r in grid)
        def calc(grid):
            s = 0
            for r in grid:
                s += sum(r)
                if s*2 == tot:
                    return True
            return False
        return calc(grid) or calc(list(zip(*grid)))

if __name__ == "__main__":
    s = Solution()
    grid = [[1,4],[2,3]]
    print(s.canPartitionGrid(grid))
